Ports/generate-available-ports.py

```diff
@@ -0,0 +1,131 @@
+    sys.exit(code)
+
+
+def join(data, separator, prefix='', postfix=''):
+    return f"{prefix}{separator.join(data)}{postfix}"
+
+
+async def read_port_properties(port):
+    if not Path(f'{port}/package.sh').exists:
+        die(f"No package for port: {port}\n", 1)
+
+    read_property_command = f"./package.sh showproperty {' '.join(PROPERTIES)}"
+    process = await asyncio.create_subprocess_shell(read_property_command, cwd=port,
+                                                    stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
+
+    print(".", file=sys.stderr, end='')
+    sys.stderr.flush()
+
+    if await process.wait() != 0:
+        error = await process.stderr.readline()
+        print('\n\n  ✖ ', file=sys.stderr)
+        sys.stderr.writelines(error.decode('utf-8'))
+
+        die(f'Could not collect package details for port: {port}  (exit code: {process.returncode})', 2)
+
+    properties = {name: (await process.stdout.readuntil(b"\n\n")).decode('utf-8').strip() for name in PROPERTIES}
+
+    if properties['port'] != port:
+        die(f"Unexpected name '{properties['port']}' for port: {port}", 3)
+
+    for name in PROPERTIES:
+        if properties[name] == '':
+            die(f"Missing property '{name}' for port: {port}", 3)
+
+    return properties
+
+
+async def read_ports():
+    ports = sorted((port.name for port in Path().iterdir() if Path(port).is_dir()), key=str.lower)
+    chunks = (ports[index:index + CONCURRENCY] for index in range(0, len(ports), CONCURRENCY))
+
+    print(f'Reading {len(ports)} ports…', file=sys.stderr)
+    properties = [await asyncio.gather(*(read_port_properties(port) for port in chunk), return_exceptions=True) for
+                  chunk in chunks]
+    print('\n', file=sys.stderr)
+    sys.stderr.flush()
+
+    return dict(zip(ports, (item for chunk in properties for item in chunk)))
+
+
+prologue = """\
+# Available ports
+
+This list is also available at [ports.serenityos.net](https://ports.serenityos.net)
+"""
+
+epilogue = """"""
+
+
+def get_settings(setting, default):
+    return dict((name, header.get(setting, default)) for name, header in HEADERS.items())
+
+
+async def main():
+    ports = await read_ports()
+
+    captions = get_settings('caption', '?')
+    min_widths = get_settings('min_width', 0)
+    formatters = get_settings('formatter', lambda x: x)
+
+    rows = [{header: formatters[header](port[header]) for header in PROPERTIES} for port in ports.values()]
+    widths = {header: max(min_widths[header], *(len(row[header]) for row in rows)) for header in PROPERTIES}
+
+    def aligned(data): return (data[header].ljust(widths[header]) for header in PROPERTIES)
+
+    def filled_with(fill): return (fill * widths[header] for header in PROPERTIES)
+
+    table_header = aligned(captions)
+    table_underline = filled_with('-')
+    table_rows = (join(aligned(row), ' | ', '| ', ' |') for row in rows)
+
+    sys.stdout.write(
```

Comment:
Two tiny nitpicks that I won't be too sad if another maintainer merges before they're addressed: 

1) sys.stdout.write --> `print()` to match w/linus's previous comment about `sys.stdout.write(file=sys.stderr)` (iirc stdout is the default for print).

2) `chdir(Path(__file__).parent / '..' / 'Ports')` 

Is this not `chdir(Path(__file__).parent)` ?

Reply:
ad 1. missed that — nicely caught! 🙇🏻‍♂️
ad 2. it is related to the second bullet in your previous comment—we need to be in that directory no matter the place we are currently, that because of relative references in the `package.sh` prologues (i.e. it is possible to call the script outside of `Ports` directory).